symlink-check/main.go

```diff
@@ -0,0 +1,285 @@
+	result := &Result{
+		FailMessages: make([]string, 0),
+	}
+
+	if len(config.Packages) > 0 {
+		for _, pkg := range config.Packages {
+			checkPackage(pkg, config.Paths, result)
+		}
+	} else {
+		for _, path := range config.Paths {
+			checkPath(path, result)
+		}
+	}
+
+	total := result.Passes + result.Fails
+	if total == 0 {
+		info("No symlinks found to check")
+	} else {
+		info(fmt.Sprintf("Tested [%d] symlinks with [%s]. [%d/%d] passed.", total, progName, result.Passes, total))
+	}
+
+	if result.Fails > 0 {
+		fmt.Println()
+		fmt.Println("FAILED SYMLINKS:")
+		for _, msg := range result.FailMessages {
+			fmt.Println(msg)
+		}
+		os.Exit(1)
+	}
+
+	os.Exit(0)
+}
+
+func isInPaths(filePath string, paths []string) bool {
+	for _, path := range paths {
+		cleanPath := filepath.Clean(path)
+		if strings.HasPrefix(filePath, cleanPath+"/") || filePath == cleanPath {
+			return true
+		}
+	}
+	return false
+}
+
+func checkPath(path string, result *Result) {
+	if _, err := os.Stat(path); err != nil {
+		if os.IsNotExist(err) {
+			result.AddFail(fmt.Sprintf("%s: Path does not exist", path))
+		} else {
+			result.AddFail(fmt.Sprintf("%s: Cannot access path: %v", path, err))
+		}
+		return
+	}
+
+	symlinks := make(chan string, 100)
+	var wg sync.WaitGroup
+
+	numWorkers := runtime.NumCPU()
+	for i := 0; i < numWorkers; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			for link := range symlinks {
+				checkSymlink(link, result)
+			}
+		}()
+	}
+
+	err := filepath.Walk(path, func(filePath string, info os.FileInfo, err error) error {
+		if err != nil {
+			return nil
+		}
+
+		if strings.HasPrefix(filePath, "/proc/") ||
+			strings.HasPrefix(filePath, "/sys/") ||
+			strings.HasPrefix(filePath, "/dev/") {
+			if info.IsDir() {
+				return filepath.SkipDir
+			}
+			return nil
+		}
+
+		if info.Mode()&os.ModeSymlink != 0 {
+			symlinks <- filePath
+		}
+
+		return nil
+	})
+
+	close(symlinks)
+	wg.Wait()
+
+	if err != nil {
+		result.AddFail(fmt.Sprintf("Error walking %s: %v", path, err))
+	}
+}
+
+func checkSymlink(link string, result *Result) {
+	target, err := os.Readlink(link)
+	if err != nil {
+		result.AddFail(fmt.Sprintf("%s: Cannot read symlink target", link))
+		return
+	}
+
+	if _, err := os.Stat(link); err != nil {
+		if os.IsNotExist(err) {
+			if target == "" {
+				result.AddFail(fmt.Sprintf("%s: Points to empty target", link))
+			} else {
+				result.AddFail(fmt.Sprintf("%s: Points to non-existent target '%s'", link, target))
+			}
+		} else {
+			result.AddFail(fmt.Sprintf("%s: Cannot access target '%s': %v", link, target, err))
+		}
+		return
+	}
+
+	if file, err := os.Open(link); err != nil {
+		result.AddFail(fmt.Sprintf("%s: Target '%s' exists but is not readable: %v", link, target, err))
+		return
+	} else {
+		file.Close()
+	}
+
+	result.AddPass(fmt.Sprintf("%s -> %s", link, target))
+}
+
+func checkPackage(pkg string, filterPaths []string, result *Result) {
+	cmd := exec.Command("apk", "info", "-eq", pkg)
```

Comment:
you depend on `apk` here so you need to add that to the pipeline dep.

Or.... you could read the database in go and _not_ depend on apk, which would be bonus points for not tainting the test environment.